myia/pipeline/steps.py

```diff
@@ -770,11 +742,19 @@ def wrapped(*args):
                 backend = NumpyChecker()
             if len(args) != len(orig_arg_t):
                 raise MyiaInputTypeError('Wrong number of arguments.')
-            args = tuple(convert_arg(arg, ot, backend) for arg, ot in
-                         zip(args, orig_arg_t))
+            args = tuple(backend.from_value(convert_arg(arg, ot), vt)
+                         for arg, ot, vt in zip(args, orig_arg_t, vm_arg_t))
             res = fn(*args)
-            res = convert_result(res, orig_out_t, vm_out_t, backend,
-                                 self.return_backend)
+            if self.return_backend:
+                if isinstance(orig_out_t, AbstractTuple):
+                    res = tuple(BackendValue(r, ot, vt, backend)
+                                for r, ot, vt in zip(res, orig_out_t.elements,
+                                                     vm_out_t.elements))
```

Comment:
Implementing `__iter__` is not impossible, but probably much slower because I would have to do IPC for each iteration. It's also not clear if the yielded elements should be `BackendValue`s or python values. And, it wouldn't work for indexing.

This is sort of clanky, but very useful.  I might add a keyword option to turn it on and off.